### PR TITLE
#437: Sanitize postal code before validation

### PR DESCRIPTION
**Pull request by tobymurray**
Fixes bug I accidentally introduced in [this PR](https://github.com/ophrescue/RescueRails/pull/657). The validation I introduced was operating on a manipulated version of the model's attribute. This sanitizes the postal code (removes whitespace and converts letters to upper case) before validation.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -256,6 +257,11 @@ def default_country_to_usa
       self.country ||= "USA"
     end
 
+    def sanitize_postal_code
+      return unless self.postal_code
+      self.postal_code = self.postal_code.delete(' ').upcase
+    end
```
Comment (jayroh) [inline app/models/user.rb]: The setter `self.postal_code=` is 👍 but there's no need here for `self`s where they are referencing the `postal_code` getters.

File: app/models/user.rb [hunk 2]
```diff
@@ -256,6 +257,11 @@ def default_country_to_usa
       self.country ||= "USA"
     end
 
+    def sanitize_postal_code
+      return unless self.postal_code
```
Comment (TheDudeWithTheThing) [inline app/models/user.rb]: `return if postal_code.blank?`

no need for self
reads better

File: app/models/user.rb [hunk 3]
```diff
@@ -256,6 +257,11 @@ def default_country_to_usa
...
+      self.postal_code = self.postal_code.delete(' ').upcase
```
Comment (TheDudeWithTheThing) [inline app/models/user.rb]: no need for self here

I didn't know about `delete`, good stuff, I think always did `tr(' ', '')`

File: app/validators/postal_code_validator.rb [hunk 4]
```diff
@@ -12,11 +12,9 @@ def validate(record)
       return
     end
 
-    postal_code = record.postal_code.delete(' ').upcase
-
-    if country.eql?(ISO3166::Country[:us]) && !postal_code.match(ZIP_CODE_REGEX)
+    if country.eql?(ISO3166::Country[:us]) && !record.postal_code.match(ZIP_CODE_REGEX)
       record.errors[:postal_code] << "should be 12345 or 12345-1234"
-    elsif country.eql?(ISO3166::Country[:ca]) && !postal_code.match(POSTAL_CODE_REGEX)
+    elsif country.eql?(ISO3166::Country[:ca]) && !record.postal_code.match(POSTAL_CODE_REGEX)
```
Comment (jayroh) [inline app/validators/postal_code_validator.rb]: What do you think about extracting these two conditionals out to more intention revealing private predicate methods?  Like `valid_us_zip_code?` and `valid_us_postal_code?` ? It's self documenting and reduces the cognitive friction a little bit more.

File: app/models/user.rb (hunk 2 above)
Comment (jayroh) [inline app/models/user.rb]: `return if postal_code.blank?` would avoid the unless.

File: app/validators/postal_code_validator.rb (hunk 4 above)
Comment (TheDudeWithTheThing) [reply app/validators/postal_code_validator.rb]: careful now, `valid_CA_postal_code?`

Comment (jayroh) [reply app/validators/postal_code_validator.rb]: Oopsies

File: app/models/user.rb (hunk 2 above)
Comment (tobymurray) [reply app/models/user.rb]: Works for me - changed in  967e880

File: app/models/user.rb (hunk 3 above)
Comment (tobymurray) [reply app/models/user.rb]: second `self` removed in 967e880. I'm not super crazy about the `delete`, as I don't find it intuitive that it removes all matches, but at least it's a word?

File: app/validators/postal_code_validator.rb (hunk 4 above)
Comment (tobymurray) [reply app/validators/postal_code_validator.rb]: Do you mean just the latter half of the conditional, or the whole conditional?

The conditional is doing two distinct things - checking what country the record is describing, and then checking the validity of the postal code in that country. I personally don't think `valid_us_zip_code` makes that more clear, and the extra layer of abstraction obscures the meat of the code from what is already what feels to me like a manageable context. Naming a function `valid_ca_postal_code` feels redundant when it's in a class called `PostalCodeValidator`.

Maybe a rename of the constants would provide a similar benefit? `VALID_CANADIAN_POSTAL_CODE` and `VALID_US_ZIP_CODE` or something?

File: app/models/user.rb (hunk 2 above)
Comment (tobymurray) [reply app/models/user.rb]: Done in 967e880

File: app/validators/postal_code_validator.rb (hunk 4 above)
Comment (markottaviani) [reply app/validators/postal_code_validator.rb]: The notion that POSTAL_CODE_REGEX = Canada and ZIP_CODE_REGEX = USA isn't obvious at least for me as a non frequent snail mailer to Canada.    So I'm a fan of renaming the Regex constants like @tobymurray mentioned.

**Pull request merged**